File: plugin/src/main/java/cn/elytra/gradle/conventions/ElytraConventionsPlugin.java
package cn.elytra.gradle.conventions;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

public class ElytraConventionsPlugin implements Plugin<Project> {

    @NotNull
    private static final String MODPACK_VERSION_DEFAULT = "2.7.4";

    @Override
    public void apply(@NotNull Project project) {
        String manifestVersion = getStringProperty(project, "elytra.manifest.version").orElse(MODPACK_VERSION_DEFAULT);
        boolean forceRefresh = getBooleanProperty(project, "elytra.manifest.no-cache").orElse(false);

        ModpackVersion.init(project, manifestVersion, forceRefresh);
    }

    private static Optional<String> getStringProperty(Project project, String propertyName) {
        Object propValue = project.findProperty(propertyName);
        if (propValue instanceof String) {
            return Optional.of((String) propValue);
        }
        return Optional.empty();
    }

    private static Optional<Boolean> getBooleanProperty(Project project, String propertyName) {
        Object propValue = project.findProperty(propertyName);
        if (propValue instanceof Boolean) {
            return Optional.of((Boolean) propValue);
        } else if (propValue instanceof String) {
            return Optional.of(Boolean.parseBoolean((String) propValue));
        }
        return Optional.empty();
    }

}

File: plugin/src/main/kotlin/cn/elytra/gradle/conventions/ElytraConventionsPlugin.kt
package cn.elytra.gradle.conventions

import cn.elytra.gradle.conventions.extension.ModpackVersionExtension
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.jvm.tasks.Jar

/**
 * The entrypoint class of Elytra Conventions Gradle Plugin.
 */
public abstract class ElytraConventionsPlugin : Plugin<Project> {
    override fun apply(project: Project) {

        // Use extension to allow lazy loading
        val extension = project.extensions.create(
            "elytraModpackVersion",
            ModpackVersionExtension::class.java
        )

        @Suppress("DEPRECATION")
        let { // run deprecated initialization here
            ModpackVersionLegacy.init(extension)
            project.extensions.extraProperties.set("elytraManifest", ModpackVersionLegacy)
        }

        // Attach manifest version to JAR
        project.afterEvaluate {
            if (extension.attachManifestVersionToJar.get()) {
                val jarTask = project.tasks.findByName("jar")
                if (jarTask is Jar) {
                    jarTask.archiveVersion.set(project.provider {
                        project.version.toString() + "+" + extension.gtnhVersion.get()
                    })
                }
            }
        }

    }

}
